### PR TITLE
mypy version bump 1.12.0

### PR DESCRIPTION
**Pull request by dirrao**
mypy version bump 1.12.0

### CODE REVIEW

**Comment by uranusjr** [comment]
We’re hitting https://github.com/python/mypy/issues/17166

**Review by uranusjr — CHANGES_REQUESTED** [review]
We’re hitting https://github.com/python/typeshed/issues/10653 (giving this a red cross to avoid accidental merge)

**Comment by potiuk** [comment]
@dirrao - are you continuing with that :) ?

**Comment by dirrao** [comment]
> @dirrao - are you continuing with that :) ?

Yes. We are waiting from the upstream to fix the bug.

**Comment by dirrao** [comment]
@uranusjr 
Are we need to adjust the existing to fix failing checks? Can please check once?

**Comment by dirrao** [comment]
@uranusjr Can you help me on the failing check?

**Comment by dirrao** [comment]
@potiuk 
We are seeing the issue related to typecasting. Do we need this explicit type casting?
```
airflow/models/dag.py:2760: error: "TaskDecoratorCollection" not callable 
[120](https://github.com/apache/airflow/actions/runs/10155935894/job/28083658202?pr=39850#step:6:121)
[misc]
[121](https://github.com/apache/airflow/actions/runs/10155935894/job/28083658202?pr=39850#step:6:122)
            return cast("TaskDecoratorCollection", functools.partial(task,...
[122](https://github.com/apache/airflow/actions/runs/10155935894/job/28083658202?pr=39850#step:6:123)
                                                   ^
[123](https://github.com/apache/airflow/actions/runs/10155935894/job/28083658202?pr=39850#step:6:124)
Found 1 error in 1 file (checked 1238 source files)
[124](https://github.com/apache/airflow/actions/runs/10155935894/job/28083658202?pr=39850#step:6:125)
Error 1 returned
```

**Comment by uranusjr** [comment]
The errore does not even make sense to me… `TaskDecoratorCollection` _is_ callable (it has `__call__` defined).

**Comment by uranusjr** [comment]
The `partial` issue looks similar to https://github.com/python/mypy/issues/17631

1.11 added special support for `partial` and that seems to break cases like this. (Not sure what the criteria are exactly.) https://github.com/python/mypy/pull/16939

**Comment by uranusjr** [comment]
Also

```
/usr/local/lib/python3.8/site-packages/apache_beam/typehints/schemas.py:486: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at [https://github.com/python/mypy/issues](https://github.com/python/mypy/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen)
version: 1.11.0
/usr/local/lib/python3.8/site-packages/apache_beam/typehints/schemas.py:486: : note: please use --show-traceback to print a traceback when reporting a bug
```

**Comment by hauntsaninja** [comment]
Try mypy 1.12 , we fixed at least one issue related to functools.partial that should help

File: hatch_build.py [hunk 1]
```diff
@@ -215,7 +215,7 @@
         # TODO: upgrade to newer versions of MyPy continuously as they are released
         # Make sure to upgrade the mypy version in update-common-sql-api-stubs in .pre-commit-config.yaml
```
Comment (hauntsaninja) [inline hatch_build.py:216]: should be updated in update-common-sql-api-stubs as per comment?

**Review by hauntsaninja — COMMENTED** [review]
Looks like there are a few things from CI that need changes

**Pull request closed without merging**